Source/GSCSTableau.m

```diff
@@ -0,0 +1,700 @@
+    CGFloat newCoefficient = [newExpression coefficientForTerm: term] * n;
+    [self addVariable: term
+         toExpression: existingExpression
+      withCoefficient: newCoefficient
+              subject: subject];
+
+    [self recordUpdatedVariable: term];
+  END_FOR_IN(newExpressionTermVariables);
+}
+
+- (void) recordUpdatedVariable: (GSCSVariable *)variable
+{
+  if ([variable isExternal])
+    {
+      [_externalParametricVariables addObject: variable];
+    }
+}
+
+- (void) setVariable: (GSCSVariable *)variable
+        onExpression: (GSCSLinearExpression *)expression
+     withCoefficient: (CGFloat)coefficient
+{
+  [expression addVariable: variable coefficient: coefficient];
+  [self recordUpdatedVariable: variable];
+}
+
+- (void) changeSubjectOnExpression: (GSCSLinearExpression *)expression
+                   existingSubject: (GSCSVariable *)existingSubject
+                        newSubject: (GSCSVariable *)newSubject
+{
+  [self setVariable: existingSubject
+       onExpression: expression
+    withCoefficient: [expression newSubject: newSubject]];
+}
+
+- (GSCSLinearExpression *) rowExpressionForVariable: (GSCSVariable *)variable
+{
+  return [_rows objectForKey: variable];
+}
+
+- (void) pivotWithEntryVariable: (GSCSVariable *)entryVariable
+                   exitVariable: (GSCSVariable *)exitVariable
+{
+// The expression "expr" represents the exit variable, which is about to be removed from the basis. 
+// This means that the old tableau should contain the equation: exitVar = expr.
+  GSCSLinearExpression *expression =
+    [self rowExpressionForVariable: exitVariable];
+  [self removeRowForVariable: exitVariable];
+
+  // Calculate an expression for the entry variable. 
+  // As "expr" has been removed from the tableau, 
+  // we can make destructive modifications to it in order to construct this expression.
+  [self changeSubjectOnExpression: expression
+                  existingSubject: exitVariable
+                       newSubject: entryVariable];
+  [self substituteOutVariable: entryVariable forExpression: expression];
+
+  if ([entryVariable isExternal])
+    {
+      [_externalParametricVariables removeObject: entryVariable];
+    }
+
+  [self addRowForVariable: entryVariable equalsExpression: expression];
+}
+
+- (NSString *) description
+{
+  NSMutableString *description =
+    [NSMutableString stringWithString: @"Tableau Information\n"];
+  [description appendFormat: @"Rows: %ld (%ld constraints)\n", [_rows count],
+                            [_rows count] - 1];
+  [description appendFormat: @"Columns: %ld\n", [_columns count]];
+  [description appendFormat: @"Infesible rows: %ld\n", [_infeasibleRows count]];
+  [description
+    appendFormat: @"External basic variables: %ld\n", [_externalRows count]];
+  [description appendFormat: @"External parametric variables: %ld\n\n",
+                            [_externalParametricVariables count]];
+
+  [description appendFormat: @"Columns: \n"];
```

Comment:
Why `appendFormat:`, this looks like a string to me.

Reply:
Fixed